Exercises/chapter-7/e1-a-robot.js

```diff
@@ -0,0 +1,194 @@
+const roads = [
+  "Alice's House-Bob's House",   "Alice's House-Cabin",
+  "Alice's House-Post Office",   "Bob's House-Town Hall",
+  "Daria's House-Ernie's House", "Daria's House-Town Hall",
+  "Ernie's House-Grete's House", "Grete's House-Farm",
+  "Grete's House-Shop",          "Marketplace-Farm",
+  "Marketplace-Post Office",     "Marketplace-Shop",
+  "Marketplace-Town Hall",       "Shop-Town Hall"
+];
+
+function buildGraph(edges) {
+  const graph = Object.create(null);
+  function addEdge(from, to) {
+    if (from in graph) {
+      graph[from].push(to);
+    } else {
+      graph[from] = [to];
+    }
+  }
+  for (let [from, to] of edges.map(r => r.split("-"))) {
+    addEdge(from, to);
+    addEdge(to, from);
+  }
+  return graph;
+}
+
+const roadGraph = buildGraph(roads);
+
+class VillageState {
+  constructor(place, parcels) {
+    this.place = place;
+    this.parcels = parcels;
+  }
+
+move(destination) {
+    if (!roadGraph[this.place].includes(destination)) {
+        return this;
+    } else {
+        const parcels = this.parcels.map(p => {
+            if (p.place !== this.place) {
+                return p;
+            }
+            return {place: destination, address: p.address};
+        }).filter(p => p.place !== p.address);
+
+        return new VillageState(destination, parcels);
+    }
+}
+
+  static random(parcelCount = 5) {
+    const parcels = [];
+    for (let i = 0; i < parcelCount; i++) {
+      const address = randomPick(Object.keys(roadGraph));
+      let place;
+      do {
+        place = randomPick(Object.keys(roadGraph));
+      } while (place === address);
+      parcels.push({place, address});
+    }
+    return new VillageState("Post Office", parcels);
+  };
+}
+
+function runRobot(state, robot, memory) {
+  for (let turn = 0;; turn++) {
+    if (state.parcels.length === 0) {
+      return turn;
+    }
+    const action = robot(state, memory);
+    state = state.move(action.direction);
+    memory = action.memory;
+
+    if (turn > 50) {
+      console.log(state);
+      return;
+    }
+    // console.log(state);
+    //console.log(`Moved to ${action.direction}`);
+  }
+}
+
+function randomPick(array) {
+  const choice = Math.floor(Math.random() * array.length);
+  return array[choice];
+}
+function findRoute(graph, from, to) {
```

Comment:
Think about how you can use line breaks to clarify your intent. What ideas go together? What ideas don't?